### PR TITLE
feat: events: version the MessageReceipt structure.

### PR DESCRIPTION
**Pull request by raulk**
This is implemented as an internal version discriminator. An alternative would've been to turn MessageReceipt into an interface with two backing structs for v0 and v1. But that approach would've broken all usage sites.

CBOR serde for MessageReceipt is no longer automatically generated. We use custom logic to deal with versioning in both directions. This logic uses the cborgen serde code as a basis.

## Checklist

Before you mark the PR ready for review, please make sure that:

- [X] Commits have a clear commit message.
- [X] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: chain/types/message_receipt_cbor.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+	"github.com/filecoin-project/go-state-types/exitcode"
+	cbg "github.com/whyrusleeping/cbor-gen"
+	"golang.org/x/xerrors"
+)
+
+// This file contains custom CBOR serde logic to deal with the new versioned
+// MessageReceipt resulting from the introduction of actor events (FIP-0049).
+
+var lengthBufMessageReceipt = []byte{132}
```
Comment (vyzo) [inline chain/types/message_receipt_cbor.go:15]: thats probably not correct, the header includes number of fields.

File: chain/types/message_receipt_cbor.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+		return err
+	}
+
+	// t.ExitCode (exitcode.ExitCode) (int64)
+	if mr.ExitCode >= 0 {
+		if err := cw.WriteMajorTypeHeader(cbg.MajUnsignedInt, uint64(mr.ExitCode)); err != nil {
+			return err
+		}
+	} else {
+		if err := cw.WriteMajorTypeHeader(cbg.MajNegativeInt, uint64(-mr.ExitCode-1)); err != nil {
```
Comment (vyzo) [inline chain/types/message_receipt_cbor.go:37]: woah, we need to be able to decode this in other languages too you know...

**Review by vyzo — COMMENTED** [review]
custom cbor looks incorrect and there is the issue of codecs in other languages, rust at minimum.

**Comment by vyzo** [comment]
Cant we do something similar to the state tree?

File: chain/types/message_receipt_cbor.go (hunk 2 above)
Comment (raulk) [reply chain/types/message_receipt_cbor.go]: This was generated by cborgen -- what is the concrete problem you see?

File: chain/types/message_receipt_cbor.go (hunk 1 above)
Comment (raulk) [reply chain/types/message_receipt_cbor.go]: Fixed.

**Review by vyzo — APPROVED** [review]
lools ok and we have test for the serializer, so green tick it is!

**Review by vyzo — COMMENTED** [review]
You could avoid the pointer wrapping and make the version specific serde free standing functions, but up to you.

**Comment by vyzo** [comment]
Note test failure.

**Comment by raulk** [comment]
> You could avoid the pointer wrapping and make the version specific serde free standing functions, but up to you.

Not worth it. I expect that to be optimized away by newtype compiler optimizations anyway.

**Pull request merged**